returnn/datasets/map.py

```diff
@@ -0,0 +1,210 @@
+"""
+Provides :class:`MapDatasetBase`
+"""
+
+from returnn.datasets.basic import DatasetSeq
+from returnn.datasets.cached2 import CachedDataset2
+from returnn.util.basic import OptionalNotImplementedError
+
+
+class MapDatasetBase(object):
+  """
+  This dataset can be used as template to implement user-side Datasets, where the data can be access in arbitrary order.
+  For global sorting, the length information needs to be known beforehand.
+  """
+
+  def __init__(self, num_outputs=None):
+    """
+    :param dict num_outputs: definition of the data as constructor parameters of a Data object
+    """
+    self.num_outputs = num_outputs or {}
+
+  def __len__(self):
+    """
+    :return: total number of sequences in the dataset
+    :rtype: int
+    """
+    raise NotImplementedError
+
+  def __getitem__(self, seq_idx):
+    """
+    This function does the actual data loading, the order can be arbitrary.
+
+    :param int seq_idx:
+    :return: The content of a single dataset entry
+    :rtype: dict[str,numpy.array]
+    """
+    raise NotImplementedError
+
+  def get_seq_len(self, seq_idx):
+    """
+    This optional function provides the sequence length for the `seq_ordering` parameter.
+    If not specified only a limited set of options is available.
+
+    :param seq_idx:
+    :return:
+    :rtype: int|None
+    """
+    raise OptionalNotImplementedError
+
+  def get_seq_tag(self, seq_idx):
+    """
+    Optionally return a sequence tag.
+
+    :param int seq_idx:
+    :return:
+    :rtype: str|None
+    """
+    return "seq-%i" % seq_idx
+
+  def get_seq_order(self, epoch=None):
+    """
+    Override to implement a dataset specific sequence order for a given epoch.
+    The number of sequences can be less than the total number.
+    This will override the effects of `partition_epoch` and `seq_ordering` when using MapDatasetWrapper.
+
+    :param epoch:
+    :return: seq_order
+    :rtype: list[int]
+    """
+    raise OptionalNotImplementedError
+
+
+class MapDatasetWrapper(CachedDataset2):
+  """
+  Takes a MapDataset and turns it into a returnn.datasets.Dataset by providing the required class methods.
+  """
+  def __init__(self, map_dataset, name=None, seq_ordering='default', random_seed_offset=None, partition_epoch=None):
```

Comment:
Why is there no `**kwargs`? Usually you always should use `kwargs`, and pass that on to `super`.
Why do you explicitly specify `name`, `seq_ordering`, etc here?